doc/usage/referencing.rst

```diff
@@ -4,83 +4,84 @@
-   section (usually ``:ref:`about-signals```).
-
-   The :rst:role:`any` role also works together with the
-   :mod:`~sphinx.ext.intersphinx` extension: when no local cross-reference is
-   found, all object types of intersphinx inventories are also searched.
+One of Sphinx's most useful features is creating automatic cross-references
+through semantic cross-referencing roles.
+A cross reference to an object description, such as ``:func:`spam```,
+will create a link to the place where ``spam()`` is documented,
+appropriate to each output format (HTML, PDF, ePUB, etc.).
+
+Sphinx supports various cross-referencing roles to create links
+to other elements in the documentation.
+In general, writing ``:role:`target``` creates a link to
+the object called *target* of the type indicated by *role*.
+The link's text depends the role but is often the same as or similar to *target*.
+
+.. _xref-modifiers:
+
+The behavior can be modified in the following ways:
+
+* **Custom link text:**
+  You can specify the link text explicitly using the same
+  notation as in reStructuredText :ref:`external links <rst-external-links>`:
+  ``:role:`custom text <target>``` will refer to *target*
+  and display *custom text* as the text of the link.
+
+* **Suppressed link:**
+  Prefixing with an exclamation mark (``!``) prevents the creation of a link
+  but otherwise keeps the visual output of the role.
+
+  For example, writing ``:py:func:`!target``` displays :py:func:`!target`,
+  with no link generated.
+
+  This is helpful for cases in which the link target does not exist;
+  e.g. changelog entries that describe removed functionality,
+  or third-party libraries that don't support :doc:`intersphinx
+  </usage/extensions/intersphinx>`.
+  Suppressing the link prevents warnings in :confval:`nitpicky` mode.
+
+* **Modified domain reference:**
+  When :ref:`referencing domain objects <ref-objects>`,
+  a tilde ``~`` prefix shortens the link text the last component of the target.
+  For example, ``:py:meth:`~queue.Queue.get``` will
+  refer to ``queue.Queue.get`` but only display ``get`` as the link text.
+
+  In HTML output, the link's ``title`` attribute
+  (that is e.g. shown as a tool-tip on mouse-hover)
+  will always be the full target name.
+
+Some of the built-in cross-reference roles are:
+
+* :rst:role:`:any: <any>`,
+  :rst:role:`:doc: <doc>`,
+  :rst:role:`:ref: <ref>`
+* :rst:role:`:confval: <confval>`,
+  :rst:role:`:envvar: <envvar>`,
+  :rst:role:`:option: <option>`
+* :rst:role:`:manpage: <manpage>`,
+  :rst:role:`:pep: <pep>`,
+  :rst:role:`:rfc: <rfc>`
+* :rst:role:`download`,
+  :rst:role:`:index: <index>`
+  :rst:role:`:numref: <numref>`,
+  :rst:role:`:keyword: <keyword>`,
+  :rst:role:`:term: <term>`,
+  :rst:role:`:token: <token>`
+* :rst:role:`!:func:`
+  (this uses the :confval:`primary_domain`, e.g. :rst:role:`:py:func: <py:func>`)
+* :ref:`Domain cross-reference roles <ref-objects>`
```

Comment:
As a short feedback, I find this listing not too helpful for the following reasons:

- The listing alomost replicates the toc of the the page.
  ![grafik](https://github.com/user-attachments/assets/0a588208-a75e-4e00-a5d5-f529975f07ae)
  The only missing items there are manpage, pep, rfc. But possibly they should also get a section on this page?
- The roles are given without context. It's while names are partially self-explanatory, you could give some category names to the bullet points.
- Stating "some" with a long list is unsettling. If possible we should make the list complete, or at the very least change the wording to "The most important ..." or similar.

Given the toc, I suggest that we can live without that list. But if you want to keep it, I suggest a bit of rework based on the above comments.

Reply:
Gone: 5eb68b2bff559838e4bac2e30b4eabb2d1929d51

The intent was to highlight the breath of roles that support cross referencing, but was perhaps ill-conceived.

A

Reply:
@timhoffm 

[typo][1]:

> (...) prefix shortens the link text the last component of the target.

should be:

> (...) prefix shortens the link text to the last component of the target.

---
---

> The roles are given without context. While names are partially self-explanatory, you could give some category names to the bullet points.

In this case less is more. I feel the TOC is repetitive by always starting with the verb *"Cross-referencing XYZ"* but the corresponding headings work well reading through the document. Adding more intro text under each heading would be a mistake because it would add clutter and waste the reader's time. (As a comparison: did you ever notice how the first half of most videos should be skipped because it just repeats an intro we already know? So lets skip the formalism of adding intros under the headings.)

---
---

> The only missing items there are manpage, pep, rfc. But possibly they should also get a section on this page?

Untangling scope overlap of the [Cross-referencing syntax][1] doc with the [Roles][2] doc should be posted separately as an issue. It seems like a complex decision.

---
---

> I'm using it in basically all of my Python module documentations as `default_role`, as recommended in the documentation, and it works great.

@mgeier We've had a number of users asking for help about ambiguous error messages caused by `:any:` when later something changes in the code (and they were using shortened syntax in the qualified name for example). Personally I always pinpoint the exact role to reduce ambiguity for myself and the reader -this also helps keep error messages more accurate- hence I don't use `default_role`.

> This makes my Python docstrings more concise

I only write cross-references in the docstring when I absolutely must.

[1]: https://www.sphinx-doc.org/en/master/usage/referencing.html#cross-referencing-syntax
[2]: https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-manpage

Reply:
@electric-coder

> > This makes my Python docstrings more concise
> 
> I only write cross-references in the docstring when I absolutely must.

Well, that's your users' loss!

My docs are full of cross-references, see e.g. https://python-sounddevice.readthedocs.io/en/0.5.0/api/convenience-functions.html

Maybe you don't like writing them because they are so clunky when specifying the exact role?
Maybe you should do your users a favor and consider using `default_role = 'any'`?

Reply:
A suitable default role depends on the type of documentation (size, how much and which kind of cross-links do you need, etc.) For example at matplotlib, we use the `py:obj` default role as a good middle-ground. We have a lot of code references and being able to write them often with just adding backticks keeps the doc sources very readable (particularly useful for docstrings that are often also used in plain text). For Python objects there is little ambiguity because the naming is determined by the code and you almost always want to have the function/class name exactly as is in the rendered docs.
We keep other references explicit, e.g. using `:doc:` or `:ref:` to prevent ambiguity. Also, by their nature, they are often longer and break the flow of reading anyway, so that adding the explicit prefixes does not make them significantly more clunky than they already are.

Reply:
> For example at matplotlib, we use the `py:obj` default role as a good middle-ground.

That is a good point, thanks for mentioning it!

When I experimented with this many years ago, `py:obj` had a severe limitation: it didn't allow me to use parentheses like this: `` `my_function()` ``. But it worked (and still works) fine with `any`.
That was the main reason why I used `any` in my documentations and I subsequently completely forgot about `py:obj` ... until now.

I just checked, and this limitation seems to have been lifted in the meantime!

Therefore, I now agree that `:any:` is a niche application and most Sphinx users should choose `default_role = 'py:obj'` in most of their (Python-based) projects!

I'm also doing that in the project I used as an example above: https://github.com/spatialaudio/python-sounddevice/pull/567